contrib/natsio/src/main/java/com/google/cloud/dataflow/contrib/natsio/NatsIO.java

```diff
@@ -0,0 +1,358 @@
+			private static final long serialVersionUID = 1L;
+			
+			private Properties props = null;
+			private Connection conn = null;
+			private String subject;
+			
+			public NatsSource(String subject, Properties props) {
+				this.subject = subject;
+				this.props = props;
+			}
+			
+			@Override
+			public NatsReader createReader(PipelineOptions options, UnboundedSource.CheckpointMark mark) {
+				return new NatsReader(subject);
+			}
+
+			@Override
+			public List<? extends UnboundedSource<KV<String,String>, UnboundedSource.CheckpointMark>> generateInitialSplits(
+					int numSplits, PipelineOptions options) throws Exception {
+				List<NatsSource> results = new ArrayList<NatsSource>();
+				// Create a NatsSource instance for each split.
+				int split = numSplits / 3;
+				for(int i = 0; i < split; i++)
+					results.add(new NatsSource(subject, props));
+				return results;
+			}
+
+			@Override
+			public Coder<UnboundedSource.CheckpointMark> getCheckpointMarkCoder() {
+				return null;
+			}
+
+			@Override
+			public Coder<KV<String,String>> getDefaultOutputCoder() {
+				return KvCoder.of(NatsIO.DEFAULT_CODER, NatsIO.DEFAULT_CODER);
+			}
+
+			@Override
+			public void validate() {
+				Preconditions.checkNotNull(props.getProperty("servers"), "need to set the 'servers' property");
+				Preconditions.checkNotNull(subject, "need to set the 'subject'");
+			}
+
+			public static class CommitMark implements UnboundedSource.CheckpointMark, Serializable {
+				private static final long serialVersionUID = 2L;
+
+				public void finalizeCheckpoint() throws IOException {}		
+			}
+				
+			private class NatsReader extends UnboundedReader<KV<String,String>> {
+				private String subject;
+				private Integer sid;
+				private ConcurrentLinkedQueue<KV<String,String>> queue;
+				private KV<String,String> current;
+				
+				public NatsReader(String subject) {
+					this.subject = subject;
+					queue = new ConcurrentLinkedQueue<KV<String,String>>();
+				}
+				
+				@Override
+				public boolean advance() throws IOException {
+					// Polling a queue to check whether an update is available in the queue.
+					current = queue.poll();
+					if (current == null)
+						return false;
+					
+					return true;
+				}
+
+				@Override
+				public UnboundedSource.CheckpointMark getCheckpointMark() {
+					return DEFAULT_MARK;
+				}
+
+				@Override
+				public byte[] getCurrentRecordId() throws NoSuchElementException {
+					return DEFAULT_ID;
+				}
+
+				@Override
+				public NatsSource getCurrentSource() {
+					return NatsSource.this;
+				}
+
+				@Override
+				public Instant getWatermark() {
+					return Instant.now();
+				}
+
+				@Override
+				public boolean start() throws IOException {
+					if (conn == null) {
+						try {
+							conn = Connection.connect(props);
+						} catch (InterruptedException e) {
+							e.printStackTrace();
+						}
+					}
```

Comment:
Since this is for a source/sink, what you want to do is effectively tell the caller above that you were closed by an interrupt. 

Throwing something like [`ClosedByInterruptException`](http://docs.oracle.com/javase/7/docs/api/java/nio/channels/ClosedByInterruptException.html) seems to make sense after cleaning up any resources.

We will fail the task, and restart the write operation again (on this or another worker).